File: app-frontend/src/api/lorebooks.ts
import { request, getBaseUrl, getHeaders, USE_MOCK, getMockDB, setMockDB } from "./config";

export interface LorebookEntry {
  keys: string[];
  content: string;
  enabled: boolean;
  constant: boolean;
  case_sensitive: boolean;
  selective: boolean;
  secondary_keys: string[];
  position: string;
  insertion_order: number;
}

export interface LorebookSummary {
  id: number;
  name: string;
  description: string;
  scan_depth?: number;
  token_budget?: number;
  recursive_scanning?: boolean;
  entries_count: number;
  created_at?: string;
}

export interface LorebookDetail extends LorebookSummary {
  entries: LorebookEntry[];
}

/**
 * 获取所有独立世界书列表
 */
export async function getLorebooks(): Promise<{ lorebooks: LorebookSummary[] }> {
  if (USE_MOCK) {
    const db = getMockDB();
    return {
      lorebooks: (db.lorebooks || []).map((lb: any) => ({
        id: lb.id,
        name: lb.name,
        description: lb.description || "",
        entries_count: (lb.entries || []).length,
      })),
    };
  }
  return request<{ lorebooks: LorebookSummary[] }>("/lorebooks");
}

/**
 * 获取单个独立世界书的详细信息
 */
export async function getLorebook(id: number): Promise<LorebookDetail> {
  if (USE_MOCK) {
    const db = getMockDB();
    const lb = (db.lorebooks || []).find((l: any) => l.id === id);
    if (!lb) throw new Error("Lorebook not found");
    return lb;
  }
  return request<LorebookDetail>(`/lorebooks/${id}`);
}

/**
 * 更新独立世界书
 */
export async function updateLorebook(
  id: number,
  data: Partial<LorebookDetail>
): Promise<{ message: string; lorebook_id: number; entries_count: number }> {
  if (USE_MOCK) {
    const db = getMockDB();
    if (!db.lorebooks) db.lorebooks = [];
    const idx = db.lorebooks.findIndex((l: any) => l.id === id);
    if (idx === -1) throw new Error("Lorebook not found");
    db.lorebooks[idx] = { ...db.lorebooks[idx], ...data };
    setMockDB(db);
    return {
      message: "Lorebook updated successfully (Mock)",
      lorebook_id: id,
      entries_count: (data.entries || []).length,
    };
  }
  return request<{ message: string; lorebook_id: number; entries_count: number }>(
    `/lorebooks/${id}`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    }
  );
}

/**
 * 删除独立世界书
 */
export async function deleteLorebook(id: number): Promise<{ message: string; lorebook_id: number }> {
  if (USE_MOCK) {
    const db = getMockDB();
    db.lorebooks = (db.lorebooks || []).filter((l: any) => l.id !== id);
    // 级联清除 MockDB 角色绑定关系
    (db.characters || []).forEach((c: any) => {
      if (c.extensions && c.extensions.lorebook_ids) {
        c.extensions.lorebook_ids = c.extensions.lorebook_ids.filter((lid: number) => lid !== id);
      }
    });
    setMockDB(db);
    return {
      message: "Lorebook deleted successfully (Mock)",
      lorebook_id: id,
    };
  }
  return request<{ message: string; lorebook_id: number }>(`/lorebooks/${id}/delete`, {
    method: "POST",
  });
}

/**
 * 上传导入世界书 JSON 文件
 */
export async function importLorebook(file: File | string): Promise<{
  message: string;
  lorebook_id: number;
  name: string;
  entries_count: number;
}> {
  if (USE_MOCK) {
    await new Promise((resolve) => setTimeout(resolve, 800));
    const db = getMockDB();
    const mockId = Date.now();
    const mockLb = {
      id: mockId,
      name: "导入的设定集 (Mock)",
      description: "Mock import description",
      entries: [
        {
          keys: ["测试"],
          content: "这是一个测试条目",
          enabled: true,
          constant: false,
          case_sensitive: false,
          selective: false,
          secondary_keys: [],
          position: "after_char",
          insertion_order: 100,
        },
      ],
    };
    if (!db.lorebooks) db.lorebooks = [];
    db.lorebooks.push(mockLb);
    setMockDB(db);
    return {
      message: "Lorebook imported successfully (Mock)",
      lorebook_id: mockId,
      name: mockLb.name,
      entries_count: mockLb.entries.length,
    };
  }

  // 1. 如果是路径字符串，采用 uni.uploadFile 上传真实的二进制文件流
  if (typeof file === "string") {
    return new Promise((resolve, reject) => {
      uni.uploadFile({
        url: `${getBaseUrl()}/lorebooks/import`,
        filePath: file,
        name: "file",
        header: getHeaders(),
        success: (res) => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            try {
              const data = JSON.parse(res.data);
              resolve(data);
            } catch (e) {
              reject(new Error(`解析响应失败: ${res.data}`));
            }
          } else {
            reject(new Error(`[importLorebook 错误 ${res.statusCode}]: ${res.data}`));
          }
        },
        fail: (err) => {
          reject(new Error(err.errMsg || "世界书导入失败"));
        },
      });
    });
  }

  // 2. 兼容 H5 网页的 File/Blob 对象
  const formData = new FormData();
  formData.append("file", file);

  const res = await fetch(`${getBaseUrl()}/lorebooks/import`, {
    method: "POST",
    headers: getHeaders(),
    body: formData,
  });

  if (!res.ok) {
    const errorBody = await res.text();
    throw new Error(`[importLorebook Error ${res.status}]: ${errorBody}`);
  }

  return res.json();
}

/**
 * 绑定角色卡到独立世界书
 */
export async function bindLorebook(characterId: number, lorebookId: number): Promise<{ message: string }> {
  if (USE_MOCK) {
    const db = getMockDB();
    const char = db.characters.find((c: any) => c.id === characterId);
    if (!char) throw new Error("Character not found");
    if (!char.extensions) char.extensions = {};
    if (!char.extensions.lorebook_ids) char.extensions.lorebook_ids = [];
    if (!char.extensions.lorebook_ids.includes(lorebookId)) {
      char.extensions.lorebook_ids.push(lorebookId);
    }
    setMockDB(db);
    return { message: "Successfully bound (Mock)" };
  }
  return request<{ message: string }>(`/lorebooks/characters/${characterId}/bind/${lorebookId}`, {
    method: "POST",
  });
}

/**
 * 解除绑定角色卡与独立世界书
 */
export async function unbindLorebook(characterId: number, lorebookId: number): Promise<{ message: string }> {
  if (USE_MOCK) {
    const db = getMockDB();
    const char = db.characters.find((c: any) => c.id === characterId);
    if (!char) throw new Error("Character not found");
    if (char.extensions && char.extensions.lorebook_ids) {
      char.extensions.lorebook_ids = char.extensions.lorebook_ids.filter((id: number) => id !== lorebookId);
    }
    setMockDB(db);
    return { message: "Successfully unbound (Mock)" };
  }
  return request<{ message: string }>(`/lorebooks/characters/${characterId}/unbind/${lorebookId}`, {
    method: "POST",
  });
}

File: app-backend/routers/lorebooks.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlalchemy.orm import Session
import json

from core.database import get_db
from core import models
from services.parse_lorebook import parse_sillytavern_lorebook
from schemas import LorebookCreate, LorebookUpdate

router = APIRouter()

@router.post("/import")
def import_lorebook(file: UploadFile = File(...), db: Session = Depends(get_db)):
    """
    导入并解析 SillyTavern 格式的世界书 JSON 文件，保存到数据库中。
    """
    try:
        content = file.file.read().decode("utf-8")
        raw_data = json.loads(content)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"文件格式错误，必须是有效的 JSON 文件: {str(e)}")

    try:
        parsed_data = parse_sillytavern_lorebook(raw_data)
        
        # 将 entries 列表序列化为 JSON 字符串存储
        entries_str = json.dumps(parsed_data["entries"], ensure_ascii=False)
        
        new_lorebook = models.Lorebook(
            name=parsed_data["name"],
            description=parsed_data["description"],
            scan_depth=parsed_data["scan_depth"],
            token_budget=parsed_data["token_budget"],
            recursive_scanning=parsed_data["recursive_scanning"],
            entries=entries_str
        )
        db.add(new_lorebook)
        db.commit()
        db.refresh(new_lorebook)
        
        return {
            "message": "Lorebook imported successfully",
            "lorebook_id": new_lorebook.id,
            "name": new_lorebook.name,
            "entries_count": len(parsed_data["entries"])
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"解析世界书失败: {str(e)}")

@router.get("")
def list_lorebooks(db: Session = Depends(get_db)):
    """
    获取系统中所有独立世界书的列表
    """
    lorebooks = db.query(models.Lorebook).all()
    result = []
    for lb in lorebooks:
        entries_count = 0
        if lb.entries:
            try:
                entries_count = len(json.loads(lb.entries))
            except Exception:
                pass
        result.append({
            "id": lb.id,
            "name": lb.name,
            "description": lb.description or "",
            "scan_depth": lb.scan_depth,
            "token_budget": lb.token_budget,
            "recursive_scanning": lb.recursive_scanning,
            "entries_count": entries_count,
            "created_at": lb.created_at.isoformat() if lb.created_at else None
        })
    return {"lorebooks": result}

@router.get("/{lorebook_id}")
def get_lorebook_detail(lorebook_id: int, db: Session = Depends(get_db)):
    """
    获取单个世界书的完整详情（包含所有条目）
    """
    lb = db.get(models.Lorebook, lorebook_id)
    if not lb:
        raise HTTPException(status_code=404, detail="Lorebook not found")
        
    entries_list = []
    if lb.entries:
        try:
            entries_list = json.loads(lb.entries)
        except Exception:
            pass
            
    return {
        "id": lb.id,
        "name": lb.name,
        "description": lb.description or "",
        "scan_depth": lb.scan_depth,
        "token_budget": lb.token_budget,
        "recursive_scanning": lb.recursive_scanning,
        "entries": entries_list,
        "created_at": lb.created_at.isoformat() if lb.created_at else None
    }

@router.put("/{lorebook_id}")
def update_lorebook(lorebook_id: int, lb_data: LorebookUpdate, db: Session = Depends(get_db)):
    """
    编辑并更新独立世界书的属性和条目
    """
    lb = db.get(models.Lorebook, lorebook_id)
    if not lb:
        raise HTTPException(status_code=404, detail="Lorebook not found")
        
    if lb_data.name is not None:
        lb.name = lb_data.name
    if lb_data.description is not None:
        lb.description = lb_data.description
    if lb_data.scan_depth is not None:
        lb.scan_depth = lb_data.scan_depth
    if lb_data.token_budget is not None:
        lb.token_budget = lb_data.token_budget
    if lb_data.recursive_scanning is not None:
        lb.recursive_scanning = lb_data.recursive_scanning
    if lb_data.entries is not None:
        # 将传入的 Pydantic 模型列表转换为字典列表后序列化为 JSON 字符串
        entries_dict_list = [entry.dict() for entry in lb_data.entries]
        lb.entries = json.dumps(entries_dict_list, ensure_ascii=False)
        
    db.commit()
    db.refresh(lb)
    
    entries_count = 0
    if lb.entries:
        try:
            entries_count = len(json.loads(lb.entries))
        except Exception:
            pass
            
    return {
        "message": "Lorebook updated successfully",
        "lorebook_id": lb.id,
        "name": lb.name,
        "entries_count": entries_count
    }

@router.delete("/{lorebook_id}")
@router.post("/{lorebook_id}/delete")
def delete_lorebook(lorebook_id: int, db: Session = Depends(get_db)):
    """
    删除指定的独立世界书，这会级联清除角色-世界书关联表中的数据。
    """
    lb = db.get(models.Lorebook, lorebook_id)
    if not lb:
        raise HTTPException(status_code=404, detail="Lorebook not found")
        
    db.delete(lb)
    db.commit()
    
    return {
        "message": "Lorebook deleted successfully",
        "lorebook_id": lorebook_id
    }

@router.post("/characters/{character_id}/bind/{lorebook_id}")
def bind_lorebook(character_id: int, lorebook_id: int, db: Session = Depends(get_db)):
    """
    将世界书绑定到指定角色卡
    """
    char = db.get(models.Character, character_id)
    if not char:
        raise HTTPException(status_code=404, detail="Character not found")
        
    lb = db.get(models.Lorebook, lorebook_id)
    if not lb:
        raise HTTPException(status_code=404, detail="Lorebook not found")
        
    if lb not in char.lorebooks:
        char.lorebooks.append(lb)
        db.commit()
        
    return {
        "message": f"Successfully bound lorebook '{lb.name}' to character '{char.name}'",
        "character_id": character_id,
        "lorebook_id": lorebook_id
    }

@router.post("/characters/{character_id}/unbind/{lorebook_id}")
def unbind_lorebook(character_id: int, lorebook_id: int, db: Session = Depends(get_db)):
    """
    将世界书与指定角色卡解绑
    """
    char = db.get(models.Character, character_id)
    if not char:
        raise HTTPException(status_code=404, detail="Character not found")
        
    lb = db.get(models.Lorebook, lorebook_id)
    if not lb:
        raise HTTPException(status_code=404, detail="Lorebook not found")
        
    if lb in char.lorebooks:
        char.lorebooks.remove(lb)
        db.commit()
        
    return {
        "message": f"Successfully unbound lorebook '{lb.name}' from character '{char.name}'",
        "character_id": character_id,
        "lorebook_id": lorebook_id
    }
